### DISCUSSION TITLE
How to add slash command dynamically?

### DISCUSSION THREAD

**Discussion by Leboweeb** [Pycord-Development/pycord#2467] [Q&A] [answered]
Suppose I want to generate a slash command from user input. I have made a function factory to return a function like so:

### Minimum Reproducible Code

```
def character_factory(character: str):
    async def inner(ctx: discord.ApplicationContext):
        """
        Command to talk to {}.
        """.format(
            character
        )
        await ctx.respond(
            f"I am a generated function. My character is {character} and I am at memory address {inner}"
        )

    return inner


@bot.slash_command(
    description="Upload characters and create slash commands for them too."
)
async def upload_characters(
    ctx: discord.ApplicationContext,
    file: discord.Attachment,
):
   characters = ["one","two","three"]
    for character in characters:
        bot.add_application_command(
            discord.SlashCommand(character_factory(character), name=character)
        )
   await ctx.respond("OK")


```

I have tried to do this and there are no errors. However, adding the command using this method simply has no effect. What do I need to do to add a command manually?

**Comment by Icebluewolf** [comment, marked as answer]
You will need to use [Bot.sync_commands](<https://docs.pycord.dev/en/stable/api/clients.html#discord.Bot.sync_commands>) passing the list of new commands. It should be noted that there is a ratelimit of 200 command registers per day. It is also possible that you want to be using guild commands. This will allow only people in the guild to use the new command (so that it can be custom per guild) and has a separate ratelimit of 200 per guild per day.

If you would like you can also join our support [Discord server](https://discord.gg/pycord)
